src/Link.cc

```diff
@@ -120,16 +93,215 @@ namespace kickcat
+            frame.setSourceMAC(src_mac);
+            int32_t written = from->write(frame.data(), to_write);
+            if (written < to_write)
+            {
+                THROW_ERROR("Can't write to interface");
+            }
+            int32_t read = to->read(frame.data(), ETH_MAX_SIZE);
+            if (read <= 0)
+            {
+                read = from->read(frame.data(), ETH_MAX_SIZE);
+                if (read <= 0)
+                {
+                    is_faulty = 1;
+                }
+            }
+            return is_faulty;
+        };
+
+        int32_t error_count = 0;
+        error_count += write_read_callback(socket_nominal_, socket_redundancy_, PRIMARY_IF_MAC);
+        error_count += write_read_callback(socket_redundancy_, socket_nominal_, SECONDARY_IF_MAC);
+
+        if (error_count > 1)
+        {
+            THROW_SYSTEM_ERROR("WriteThenRead was not able to read anything on both interfaces");
+        }
+
+        if (frame.ethernet()->type != ETH_ETHERCAT_TYPE)
+        {
+            THROW_ERROR("Invalid frame type");
+        }
+
+        int32_t current_size = frame.header()->len + sizeof(EthernetHeader) + sizeof(EthercatHeader);
+
+        // Take padding into account.
+        if (current_size < ETH_MIN_SIZE)
+        {
+            current_size = ETH_MIN_SIZE;
+        }
+        if (current_size != to_write)
+        {
+            THROW_ERROR("Wrong number of bytes read");
+        }
+        frame.setIsDatagramAvailable();
+    }
+
+
+    void Link::sendFrame()
+    {
+        // save number of datagrams in the frame to handle send error properly if any
+        int32_t const datagrams = frame_nominal_.datagramCounter();
+
+        bool is_frame_sent_nominal = true;
+        frame_nominal_.setSourceMAC(PRIMARY_IF_MAC);
+        int32_t toWrite = frame_nominal_.finalize();
+        int32_t written = socket_nominal_->write(frame_nominal_.data(), toWrite);
+        if (written < 0)
+        {
+            is_frame_sent_nominal = false;
+            DEBUG_PRINT("Nominal: write failed\n");
+        }
+        else if (written != toWrite)
+        {
+            is_frame_sent_nominal = false;
+            DEBUG_PRINT("Nominal: Wrong number of bytes written");
+        }
+
+        bool is_frame_sent_redundancy = true;
+        frame_nominal_.setSourceMAC(SECONDARY_IF_MAC);
+        written = socket_redundancy_->write(frame_nominal_.data(), toWrite);
+        frame_nominal_.clear();
+        frame_redundancy_.clear();
+        frame_redundancy_.resetContext();
+
+        if (written < 0)
+        {
+            is_frame_sent_redundancy = false;
+            DEBUG_PRINT("Redundancy: write failed\n");
+        }
+
+        else if (written != toWrite)
+        {
+            is_frame_sent_redundancy = false;
+            DEBUG_PRINT("Redundancy: Wrong number of bytes written");
+        }
+
+        if (is_frame_sent_nominal or is_frame_sent_redundancy)
+        {
+            sent_frame_++;
```

Comment:
Better use prefix increment (++sent_frame_) instead since you'll never need the old value.

Reply:
Why is it better ?

Reply:
i = 2
b = ++i
a = i++

=> a == b == 3
By convention, I prefer to use prefix increment when the value before increment is of no use.